src/gl.rs

```diff
@@ -518,6 +518,447 @@ pub trait Gl {
     fn get_debug_messages(&self) -> Vec<DebugMessage>;
 }
 
+struct ErrorCheckingGl<T: Gl> {
+    gl: T,
+}
+
+impl<T: Gl + 'static> ErrorCheckingGl<T> {
+    fn maybe_wrap(fns: T) -> Rc<Gl> {
+        if cfg!(feature = "assert_on_error") {
+            Rc::new(ErrorCheckingGl { gl: fns }) as Rc<Gl>
+        } else {
+            Rc::new(fns) as Rc<Gl>
+        }
+    }
+}
+
+macro_rules! gl_apis {
+    ($(fn $name:ident(&self$(, $arg:ident: $t:ty)*)$( -> $retty:ty)*;)+) => {
+        $(fn $name(&self$(, $arg:$t)*)$( -> $retty)* {
+            let rv = self.gl.$name($($arg,)*);
+            assert_eq!(self.gl.get_error(), 0);
+            rv
+        })+
+    }
+}
+
+impl<T: Gl> Gl for ErrorCheckingGl<T> {
+    gl_apis! {
+        fn get_type(&self) -> GlType;
```

Comment:
Is it possible to re-use this list of signatures for the `trait Gl` to avoid repeating ourselves?

Reply:
I extended the macro to take the entire trait declaration as input.